### ISSUE TITLE
Bug: Using Sourcegraph.com GraphQL API from other websites is broken

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#18847]
Problem: If you try to use `https://sourcegraph.com/.api/graphql` from another website, it is blocked due to CORS ~because we're not setting any `Content-Security-Policy` for responses from that URL, it defaults and is thus blocked:~

~Content Security Policy: The page’s settings blocked the loading of a resource at https://sourcegraph.com/.api/graphql?SearchContexts (“default-src”).~

**Are you ready for some history?**

* Based on my memory, we have always intended for the Sourcegraph.com GraphQL API to be used as broadly and accessibly as possible - including by other websites, by unauthenticated users on https://sourcegraph.com/api/console, via the `src` CLI, and literally everywhere else.
* We explicitly designed the CORS handling of the GraphQL API to _only ever allow cookie-based session auth_ [_if on the same domain_ (in so-called "non-simple" CORS requests)](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/session/session.go#L299-337)
* This is closely related to this old issue I have open about [removing our redundant CSRF cookies #7658](https://github.com/sourcegraph/sourcegraph/issues/7658) and if you're wondering "how is that secure?" see [my detailed write-up here back in 2018](https://github.com/sourcegraph/sourcegraph/issues/227#issuecomment-426482380) which is still true today - or this more brief explanation of [OWASP: using custom request headers to prevent CSRF](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#use-of-custom-request-headers).

In short:

1. ~https://sourcegraph.com/.api/graphql should have a `Content-Security-Policy` which allows requests from any origin.~
2. ~This should be completely safe and secure to enable, and was the original intended behavior - but **obviously** needs verification.~

**Comment by felixfbecker** [comment]
Wait, isn't the issue here the _page's_ CSP? Or why would sourcegraph.com/.api/graphql's CSP cause itself to be blocked? Missing CORS headers on our end would cause that, but not CSP, afaik. Unless the _page_ defined _in its own CSP_ that sourcegraph.com is not an allowed fetch target.

**Comment by slimsag** [comment]
@felixfbecker yeah, seems you're right - not gonna lie, I was super tired when I wrote this issue up super quickly - I should've emphasized this part a lot more:

> but obviously needs verification.

**Comment by conwnet** [comment]
It seems that the CORS check is used to prevent the CSRF attacks currently.

The crosss-site cookie will only be carried when the CORS header `Access-Control-Allow-Origin` is not a wildcard value `*`. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

So in my opinion, we can set the `Access-Control-Allow-Origin` header with a `<origin>` value explicitly only if we need the request carry the cross-site cookie. Otherwise, we can easily set it with the wildcard value `*` for other cross-site request. 😃

Hope this infomation can make help with you.

**Comment by slimsag** [comment]
Following up on this here: https://github.com/sourcegraph/sourcegraph/pull/23949

**Comment by slimsag** [comment]
PR to fix this and enable public access of the Sourcegraph.com API is out: #27937
